### PR TITLE
Add video/audio extension awareness

### PR DESCRIPTION
**Pull request by obenland**
Adds a widget-specific way to determine whether they have content or not and populates all extensions based on their values.

### CODE REVIEW

**Comment by obenland** [comment]
One thing missing here is to populate those extensions from the PHP side on first load

File: wp-admin/js/widgets/media-widgets.js [hunk 1]
```diff
@@ -713,6 +713,20 @@ wp.mediaWidgets = ( function( $ ) {
...
+			// Reset all video/audio sources.
+			wp.media.view.settings.embedExts.map( function( ext ) {
+
+				// Only reset when the main source changes.
+				if ( ! ( ext in modelProps ) || isEmbed ) {
+					modelProps[ ext ] = '';
+				}
+			} );
+
+			extension = mediaFrameProps.url.split( '.' ).pop();
```
Comment (timmyc) [inline wp-admin/js/widgets/media-widgets.js]: Do we need to be concerned about case here, i.e. if a file has an extension of `.MP4` instead of `.mp4` maybe call `toLowerCase()` on it just to be safe?

File: wp-admin/js/widgets/media-widgets.js [hunk 2]
```diff
@@ -713,6 +713,20 @@ wp.mediaWidgets = ( function( $ ) {
...
+			modelProps[ extension ] = mediaFrameProps.url;
```
Comment (timmyc) [inline wp-admin/js/widgets/media-widgets.js]: Should we check to see if `wp.media.view.settings.embedExts[ extension ]` exists first?  As-is won't this set a prop of `jpg` on image widgets?

File: tests/qunit/test-media-image-widget.js [hunk 3]
```diff
@@ -62,9 +62,9 @@
 		// Test mapMediaToModelProps
 		mappedProps = imageWidgetControlInstance.mapMediaToModelProps( { link: 'file', url: testImageUrl } );
 		equal( mappedProps.link_url, testImageUrl, 'mapMediaToModelProps should set file link_url according to mediaFrameProps.link' );
-		mappedProps = imageWidgetControlInstance.mapMediaToModelProps( { link: 'post', postUrl: 'https://wordpress.org/image-2/' } );
+		mappedProps = imageWidgetControlInstance.mapMediaToModelProps( { link: 'post', postUrl: 'https://wordpress.org/image-2/', url: testImageUrl  } );
 		equal( mappedProps.link_url, 'https://wordpress.org/image-2/', 'mapMediaToModelProps should set file link_url according to mediaFrameProps.link' );
-		mappedProps = imageWidgetControlInstance.mapMediaToModelProps( { link: 'custom', linkUrl: 'https://wordpress.org' } );
+		mappedProps = imageWidgetControlInstance.mapMediaToModelProps( { link: 'custom', linkUrl: 'https://wordpress.org', url: testImageUrl  } );
```
Comment (timmyc) [inline tests/qunit/test-media-image-widget.js]: Should we add some more tests here to cover the extensions being set/reset?

**Review by timmyc — COMMENTED** [review]
I used the steps outlined in #126 and verified that the alternate sources are no longer showed when switching from a local file -> YouTube video.  Few other minor comments.

File: wp-includes/widgets/class-wp-widget-media.php [hunk 4]
```diff
@@ -420,4 +402,38 @@ public function render_control_template_scripts() {
...
+	 * Creates and returns a link for an attachment.
+	 *
+	 * @since 4.8.0
+	 * @access protected
+	 *
+	 * @param WP_Post $attachment Attachment object.
+	 * @param string  $type       link type.
+	 * @return string
+	 */
+	protected function create_link_for( $attachment, $type = '' ) {
```
Comment (westonruter) [inline wp-includes/widgets/class-wp-widget-media.php]: I think this method can be just removed entirely, right? It's not even used.

File: wp-admin/js/widgets/media-widgets.js (hunk 2 above)
Comment (westonruter) [inline wp-admin/js/widgets/media-widgets.js]: Shouldn't this only be done if `isEmbed`? Also, if `! isEmbed` then I believe that `modelProps.url = ''` needs to be done, so that we reserve `url` for external embeds and then the extension props for media uploaded to the media library?

Comment (obenland) [reply wp-admin/js/widgets/media-widgets.js]: It would, but the schema takes care of that for us and only adopts whitelisted values

File: tests/qunit/test-media-image-widget.js (hunk 3 above)
Comment (obenland) [reply tests/qunit/test-media-image-widget.js]: Absolutely we should

File: wp-includes/widgets/class-wp-widget-media.php (hunk 4 above)
Comment (obenland) [reply wp-includes/widgets/class-wp-widget-media.php]: Ha! Even better. I didn't check

File: wp-admin/js/widgets/media-widgets.js (hunk 2 above)
Comment (obenland) [reply wp-admin/js/widgets/media-widgets.js]: If we do it for uploaded files as well, we'll automatically populate the additional sources part in the edit modal.

I'm not sure I understand why you want to move away from uploaded files using `url`. Should that be set no matter what? When would we encounter an attachment without it?

Comment (westonruter) [reply wp-admin/js/widgets/media-widgets.js]: @obenland So you're fine with there being a `url` and `mp4` props that have the same value?

Comment (obenland) [reply wp-admin/js/widgets/media-widgets.js]: Unless it introduces unexpected behavior or errors, I don't see what the harm is. And we can continue to rely on `url` in all cases

Comment (westonruter) [reply wp-admin/js/widgets/media-widgets.js]: OK, sure. And I guess there is a slight advantage there because it allows the user to indicate the “primary” source in a sense.

File: wp-admin/js/widgets/media-widgets.js [hunk 5]
```diff
@@ -713,6 +713,20 @@ wp.mediaWidgets = ( function( $ ) {
...
+			wp.media.view.settings.embedExts.map( function( ext ) {
+
+				// Only reset when the main source changes.
+				if ( ! ( ext in modelProps ) || isEmbed ) {
+					modelProps[ ext ] = '';
+				}
+			} );
+
+			extension = mediaFrameProps.url.split( '.' ).pop().toLowerCase();
+			modelProps[ extension ] = mediaFrameProps.url;
```
Comment (westonruter) [inline wp-admin/js/widgets/media-widgets.js]: Doesn't this need to be done only if `! isEmbed`? In other words, if I select a YouTube video, then `extension` will not be valid.

File: wp-includes/widgets/class-wp-widget-media.php [hunk 6]
```diff
@@ -420,4 +402,17 @@ public function render_control_template_scripts() {
...
 	}
+
+	/**
+	 * Whether the widget has content to show.
+	 *
+	 * @since 4.8.0
+	 * @access protected
+	 *
+	 * @param array $instance Widget instance props.
+	 * @return bool
```
Comment (westonruter) [inline wp-includes/widgets/class-wp-widget-media.php]: Add return description.

File: wp-includes/widgets/class-wp-widget-media.php [hunk 7]
```diff
@@ -420,4 +402,17 @@ public function render_control_template_scripts() {
...
+	 */
+	protected function has_content( $instance ) {
+		return ( $instance['attachment_id'] && 'attachment' !== get_post_type( $instance['attachment_id'] ) ) || $instance['url'];
```
Comment (westonruter) [inline wp-includes/widgets/class-wp-widget-media.php]: Not sure I understand the logic in this. Shouldn't it include `'attachment' === get_post_type( $instance['attachment_id'] )` (`===` not `!==`)?

Comment (obenland) [reply wp-includes/widgets/class-wp-widget-media.php]: Yes, good catch! Missed that one in my conversion

File: wp-admin/js/widgets/media-widgets.js (hunk 5 above)
Comment (obenland) [reply wp-admin/js/widgets/media-widgets.js]: Oy, I've only tested it with file urls so far

Comment (obenland) [reply wp-admin/js/widgets/media-widgets.js]: If `extension` is not valid it will be purged by the instance schema. 

Do you want us to be more verbose here and test against an extension whitelist?

Comment (westonruter) [reply wp-admin/js/widgets/media-widgets.js]: I'm more interested in the case where `mediaFrameProps.url` is equal to `https://www.youtube.com/watch?v=HTgYHHKs0Zw`, as in that case the “extension” will be `com/watch?v=HTgYHHKs0Zw`. So yeah, I guess we should skip this entirely if it doesn't have a file extension. Also, I think we should perhaps explicitly strip off any query or hash fragments from the URL so that if someone supplies a file URL with any additional params, the extension will get picked up properly.

In other words, something like:

```js
extension = mediaFrameProps.url.replace( /#.*$/, '' ).replace( /\?.*$/, '' ).split( '.' ).pop().toLowerCase()
```

And then yes, I think it would probably be good form to check if `extension` exists among `control.model.schema`.

**Comment by timmyc** [comment]
Query string removal, extension check, and a few more tests added in fb597c8

**Comment by westonruter** [comment]
It seemed to make more sense to reset the media source props in `getModelPropsFromMediaFrame` when we know that a _new_ media item was selected. Otherwise, `mapMediaToModelProps` shouldn't be resetting props when returning from the edit frame. I did this in fe9d590 though sadly this means some of the new tests can't still apply at `getModelPropsFromMediaFrame` is not testable since it relies on the media manager JS.

**Review by westonruter — APPROVED** [review]
